processes/addFulfillManifest.py

```diff
@@ -0,0 +1,138 @@
+            self.updateManifest()
+        elif self.process == 'custom':
+            timeStamp = self.ingestPeriod
+            startTimeStamp = datetime.strptime(timeStamp, '%Y-%m-%dT%H:%M:%S')
+            self.updateManifest(startTimeStamp)
+
+    def updateManifest(self, startTimeStamp=None):
+
+        '''Load batch of LA works from past 24 hours'''
+
+        batches = self.load_batch()
+        for batch in batches:
+            for c in batch['Contents']:
+                if c["LastModified"] > startTimeStamp.replace(tzinfo=timezone.utc):
+                    currKey = c['Key']
+                    metadataObject = self.getObjectFromBucket(currKey, self.s3Bucket)
+                    self.update_batch(metadataObject, self.s3Bucket, currKey)
+
+    def update_batch(self, metadataObject, bucketName, currKey):
+        dbManager = DBManager(
+        user=os.environ.get('POSTGRES_USER', None),
+        pswd=os.environ.get('POSTGRES_PSWD', None),
+        host=os.environ.get('POSTGRES_HOST', None),
+        port=os.environ.get('POSTGRES_PORT', None),
+        db=os.environ.get('POSTGRES_NAME', None)
+    )
+
+        dbManager.generateEngine()
+
+        dbManager.createSession()
+
+        metadataJSON = json.loads(metadataObject['Body'].read().decode("utf-8"))
+        metadataJSONCopy = copy.deepcopy(metadataJSON)
+        
+        metadataJSON = self.linkFulfill(metadataJSON, dbManager)
+        metadataJSON = self.readingOrderFulfill(metadataJSON, dbManager)
+        metadataJSON = self.resourceFulfill(metadataJSON, dbManager)
+        metadataJSON = self.tocFulfill(metadataJSON, dbManager)
+
+        if metadataJSON != metadataJSONCopy:
+            try:
+                fulfillManifest = json.dumps(metadataJSON, ensure_ascii = False)
+                return self.putObjectInBucket(
+                    fulfillManifest, currKey, bucketName
+                )
+            except ClientError as e:
+                logging.error(e)
+
+    def linkFulfill(self, metadataJSON, dbManager):
+        for i in metadataJSON['links']:
+            i['href'] = self.fulfillReplace(i, dbManager)
+
+        return metadataJSON
+            
+    def readingOrderFulfill(self, metadataJSON, dbManager):
+        for i in metadataJSON['readingOrder']:
+            i['href'] = self.fulfillReplace(i, dbManager)
+
+        return metadataJSON
+
+    def resourceFulfill(self, metadataJSON, dbManager):
+        for i in metadataJSON['resources']:
+            i['href'] = self.fulfillReplace(i, dbManager)
+
+        return metadataJSON
+
+    def tocFulfill(self, metadataJSON, dbManager): 
+        for i in metadataJSON['toc']:
+            if 'pdf' in i['href'] \
+                or 'epub' in i['href']:
+                    for link in dbManager.session.query(Link) \
+                        .filter(Link.url == i['href'].replace('https://', '')):
+                            i['href'] = f'http://127.0.0.1:5050/fulfill/{link.id}'
```

Comment:
Hi Dmitri, thanks for your work on this! This script will run on QA and production in order to add manifests on those servers in the future, right? If that is the case, I don’t think a hard coded URL to the dev environment will work since the processes will run remotely. I had to deal with the same issue of dynamically guiding the urls in many of the endpoints when adding the fulfill endpoints to the JSON (e.g. the work endpoint).

Reply:
Appreciate the feedback Lyndsey. You're right on a hard coded url not working in this case, so I'll update these urls to be more dynamic for the prod and QA environments.